### DISCUSSION TITLE
New streak behaviour doesn't work for all use cases

### DISCUSSION THREAD

**Discussion by keysmashes** [iSoron/uhabits#560] [Feature Requests]
I track a habit which I can only do once every 2 days; this used to work fine, but in the 1.8 beta (since #124 was fixed) the reminders now aren't scheduled properly if I miss a single day.

For example:

- day 1: complete the task
- day 2: do not complete the task
- day 3: do not complete the task
- day 4: complete the task
- day 5: do not complete the task

Under the old system, day 3 would be uncompleted (day 5 would be completed instead), which is for this task the correct behaviour. Under the new system, when I complete the task on day 4, day 3 also gets completed, and I get a reminder to do the task on day 5 – even though I can't do it yet, because I did it yesterday.

I don't really have a good idea for a solution to this, sorry. Is it possible to implement an option to switch between the old and new streak calculation for each task?

**Comment by chennemann** [comment]
Hey @sersorrel,

could you maybe provide a specific example to when this pattern is that relevant? I cannot think of anything specific so it's hard to justify having both approaches which would come with a significant cost in regards to usability and code complexity.

**Comment by keysmashes** [comment]
It's not a great example, but: I guess if you imagine that the app supported completing a task multiple times per day, then imagine you have a task to brush your teeth twice a day – it wouldn't make sense to brush your teeth three times in a day after you forgot to brush them once on the previous day. Generally it would be useful for any other task where the need to complete it increases over time, but each time you complete it the need to complete it again drops to zero, so probably other stuff related to cleaning or topping up fuel or changing batteries or something like that.

It's not a big deal if it's too much work and complexity to keep both algorithms, I can just keep using the old version (or compile my own, if I can find the commit where the new algorithm was added).

**Comment by iSoron** [comment]
Loop 1.8.8 (available soon for beta testers) has yet another algorithm, which should work well both with the example you show here, as well as the one from #124. Please try it out when it becomes available to you, and let me know if it behaves as you expect.

Here is a screenshot for "1 time every 2 days" habits:
![Screenshot_20200621-141221](https://user-images.githubusercontent.com/4699657/85233135-cef23680-b3c9-11ea-93ee-0e7490c6cbd5.png)

~~Here is a screenshot for "3 times every 7 days":~~ See below for a fixed example.
![Screenshot_20200621-141324](https://user-images.githubusercontent.com/4699657/85233141-db768f00-b3c9-11ea-84e3-f8674b029e64.png)

**Comment by myrdd** [comment]
> Here is a screenshot for "3 times every 7 days":
> ![Screenshot_20200621-141324](https://user-images.githubusercontent.com/4699657/85233141-db768f00-b3c9-11ea-84e3-f8674b029e64.png)

This somehow looks more like "2 times every 7 days" ... are you sure it's 3 times @iSoron?
if it's indeed 3 times, why is 24.-28.(March) and 17./18. (May) marked?

I've got quite the same use case as @sersorrel, but "1 time every 7 days".
Or, better said, **"max. 7 days between one completion and the next one"**.

---

PS: Thanks for the great app!!

( PPS: FWIW, I once have been using HabitRPG, and they had (or still have) a similar issue..
This comment is quite interesting https://github.com/HabitRPG/habitica/issues/6941#issuecomment-326780411 )

**Comment by iSoron** [comment]
@myrdd You're right; my example was wrong. Below is a fixed example, "2 times per week". Also, thanks for the link. Very interesting to see how other apps deal with this. 

![Screenshot_20200622-194427](https://user-images.githubusercontent.com/4699657/85348453-43f46780-b4c1-11ea-9b2b-8e65f6c7363c.png)

**Comment by keysmashes** [comment]
It'll take me a few days to see how notifications behave now, but from how my history looks this looks like exactly the behaviour I was hoping for. Thank you so much!

**Comment by myrdd** [comment]
I just tested 1.8.8 and the "1 time every 7 days" case works exactly how I would have expected. So indeed, thank you so much!

FTR, here's a screenshot of my use case:
![Screenshot_20200624-075536](https://user-images.githubusercontent.com/4286829/85506817-a09a7400-b5e0-11ea-8076-3ecc735afbf2.jpg)

Still, @iSoron, I'm wondering why in your screenshot (I cite below) days 24 to 28 of march are marked. I tried to reproduce the result of your screenshot, using v1.8.8, but 24–28 appear gray. (Well, that's what I would have expected. Just wondering.)

> ![Screenshot_20200622-194427](https://user-images.githubusercontent.com/4699657/85348453-43f46780-b4c1-11ea-9b2b-8e65f6c7363c.png)

**Comment by KristianTashkov** [comment]
There are some inconsistencies created by the current logic and I'm not exactly sure how to solve for them:
1) Very similar example as the one above, but producing a different result:
![image](https://user-images.githubusercontent.com/3768676/88102056-9e0f4800-cba7-11ea-9ab9-2e8d2ff6bcac.png)
The two sides of this example are: you technically did do habit 2 types per week, but the second repetition goes to next week instead of the current one. But if you move the attribution to the previous week making two full weeks, this means that next Sunday (the 14th) this habit will have a reminder, even though you just did it twice in a roll. Not sure what's the correct interpretation there.
2) The current logic has a huge flaw with double counting repetitions. For a twice a week habit this is possible:
![image](https://user-images.githubusercontent.com/3768676/88102310-03fbcf80-cba8-11ea-8bc6-53c42e0ef9aa.png)
Even though I was supposed to do the habit twice per week, doing it only one counts for both the previous and the next week.

**Comment by myrdd** [comment]
IMHO it should be possible to distinguish between
* x times per calendar week (@KristianTashkov's case)
* x times within 7 days (any 7d series, not bound to calendar weeks; that's my use case from above)

@sersorrel's use case would fall into the first type, but with 2d groups instead of 7d ”calendar week“ groups.
No idea how to support both types without sacrificing too much usability and simplicity.. Personally I'd be satisfied with some hidden ”advanced“ setting.

**Comment by KristianTashkov** [comment]
I think I figured out what is happening. When the habit must be done more than once a week, we are heavily double counting days. If we look at the staircase example I gave above, this is a 2 times per week habit, but if you change it to 1 times per week, you get the same graph for the middle weeks, which intuitively sounds wrong.  If we stop double counting this is what the staircase example looks like with 2 times per week:
![image](https://user-images.githubusercontent.com/3768676/88262076-31439d00-ccd0-11ea-9ac4-f880c3ed0c93.png)
And for @iSoron example it looks like this with aligns more with @myrdd understanding of how 1 per week should work, just scaled to 2 per week (so two repetitions after another instead of 1):
![image](https://user-images.githubusercontent.com/3768676/88262159-546e4c80-ccd0-11ea-9aef-cb891bb83fb3.png)

While this fixes the double counting where 2 repetitions are counted for both the previous week and the next week, see example of current behavior:
![image](https://user-images.githubusercontent.com/3768676/88262545-132a6c80-ccd1-11ea-8686-6bfcf3d5d74c.png)
It doesn't solve for the use case of "I have to work out 2 times per week, regardless of which days":
![image](https://user-images.githubusercontent.com/3768676/88263440-b465f280-ccd2-11ea-87b1-aa8d448f6e43.png)
To get away from the concept of "calendar days" because I don't think this is relevant in this case, it could be reformulated as "if you look at two successful repetition of the habit (in this case 4 reps in 2 weeks), if you managed to do it, then the whole interval is considered successful". This could be an option for the habit since it clashes with the concept of @myrdd where a habit MUST be done every X days or it is considered failed. This new option only makes sense for habits that you can retroactively "correct", even if that makes you do more repetitions in the period you selected to compensate. For example weekly "working out" habit that you usually do on Mondays. If you miss a Monday, but then manage to do a workout on Wednesday and afterwards continue doing it every Monday. Even though you missed 2 days, you compensated by having a shorter rest of 2 days so it all events out.

**Comment by iSoron** [comment]
> There are some inconsistencies created by the current logic and I'm not exactly sure how to solve for them:

For each one of these examples, could you let us know what did you expect the result to be?

>  If we stop double counting this is what the staircase example looks like with 2 times per week:

How did you produce this chart? Was it manually or did you modify the app in some way? 

> The current logic has a huge flaw with double counting repetitions.

Until I understand these examples better, I am erring on the side of caution and giving the user more "off" days than they potentially deserve. In the chart where you stop double counting, why have you decided to mark July 12-18 as "not completed", even though the user did complete the habit twice during that week? I understand that you decided to pair July 12 with the previous checkmark, instead of the following one, but why?

**Comment by KristianTashkov** [comment]
> How did you produce this chart? Was it manually or did you modify the app in some way?

I modified the app, when we are building intervals and you make a successful one, continue the logic from the repetition after the last one used. This avoids the double counting I mention.
> In the chart where you stop double counting, why have you decided to mark July 12-18 as "not completed", even though the user did complete the habit twice during that week?

I guess it depends on your definition of "two times per week" habit. In my mind if I hear that someone is doing something 2 times per week, I expect that after a year (52) weeks, they would have done that habit at least 104 times. In the current state of the app, you can get away with doing it 5 times less every 7 weeks (the staircase example I showed with the current behavior) for an overall minimum of around 70 times for the app to show the entire year as "completed".
> For each one of these examples, could you let us know what did you expect the result to be?

It's hard for me because it really depends on the habit and how we implement "X times every Y days". Are these Y days overlapping or not? @myrdd's use case is different than mine. For most of my habits "2 times per 5 days" means that it doesn't matter which days exactly I do the habit, I can even do it in four consecutive days, but as long as I have the timeline split somehow into 5 day chunks, and for each of them I have at least two repetitions, that means I was successful. Right now if you have enough consecutive repetitions to abuse the double counting behavior (`snapIntervals` pulling the beginning the most it can backwards) you basically end up in a situation that 2 consecutive days mark the previous 3 and the next 3 days as "successful" at the same time.

**Comment by KristianTashkov** [comment]
As an example of the different behaviors. This is 1 per week in the app:
![image](https://user-images.githubusercontent.com/3768676/88461515-da46ef00-ceac-11ea-9af5-e52c56f5e611.png)
For my use case it makes sense to be:
![image](https://user-images.githubusercontent.com/3768676/88461523-e3d05700-ceac-11ea-9152-cb0faca779fc.png)
Both are valid but are useful for different types of habits, which is why I think we should somehow support both with a setting.
For the second behavior there are two ways to explain it:
1) In each 7 week intervals Sunday -> Saturday I did it once so the whole period should be marked. Then the next week I completely missed it, and the next repetition starts a new 7 week period.
2) If you look at the first two repetitions, since you have `2 * required` repetitions in `2 * frequency` period, then the whole 2 week period should be marked as completed.

It only makes sense for habits that you can do retroactively to catch up with you missing it earlier in the week, e.g I want to work out 1 time per week, but I don't care which day. 

Although this discussion is entirely independent of the decision whether more than 1 times per X habits should allow their ends to be double counted in two separate X periods.
